snapshot/snapshotfs/upload_test.go

```diff
@@ -692,6 +692,136 @@ func TestUpload_VirtualDirectoryWithStreamingFile(t *testing.T) {
 	}
 }
 
+func TestUpload_StreamingDirectory(t *testing.T) {
+	cases := map[string]bool{
+		"MultipleIterations": true,
+		"SingleIteration":    false,
+	}
+
+	for desc, iterations := range cases {
+		t.Run(desc, func(t *testing.T) {
+			ctx := testlogging.Context(t)
+			th := newUploadTestHarness(ctx, t)
+
+			defer th.cleanup()
+
+			t.Logf("Uploading streaming directory with mock file")
+
+			u := NewUploader(th.repo)
+
+			policyTree := policy.BuildTree(nil, policy.DefaultPolicy)
+
+			files := []fs.Entry{
+				mockfs.NewFile("f1", []byte{1, 2, 3}, defaultPermissions),
+			}
+
+			staticRoot := virtualfs.NewStaticDirectory("rootdir", []fs.Entry{
+				virtualfs.NewStreamingDirectory(
+					"stream-directory",
+					func(innerCtx context.Context, callback func(context.Context, fs.Entry) error) error {
+						for _, f := range files {
+							if err := callback(innerCtx, f); err != nil {
+								return err
+							}
+						}
+
+						return nil
+					},
+					iterations,
+				),
+			})
+
```

Comment:
In this test, where are we testing the multiple or single pass ? Looks like we are testing dynamic adding of a directory to root dir and adding a file to the sub dir. Do we need to have the for loop for multiple and single ReadDir/iternation ?

Reply:
The test was intended to check both when the `StreamingDirectory` supports iterating multiple times and when it does not. The estimator in `uploader` also reads through all entries of a directory, so this was meant to check that it won't break. However, that test case has a race condition, so I've also added a separate test in `estimate_test.go`